core/services/pipeline/orm.go

```diff
@@ -0,0 +1,360 @@
+			done = true
+			return nil
+		} else if err != nil {
+			return errors.Wrap(err, "error finding next task run")
+		}
+
+		// Fill in the TaskSpec
+		err = tx.Where("id = ?", ptRun.PipelineTaskSpecID).First(&ptRun.PipelineTaskSpec).Error
+		if gorm.IsRecordNotFoundError(err) {
+			// done = true
+			return err
+		} else if err != nil {
+			return errors.Wrap(err, "error finding next task run's spec")
+		}
+
+		// Fill in the PipelineRun
+		err = tx.Where("id = ?", ptRun.PipelineRunID).First(&ptRun.PipelineRun).Error
+		if gorm.IsRecordNotFoundError(err) {
+			// done = true
+			return err
+		} else if err != nil {
+			return errors.Wrap(err, "error finding next task run's pipeline.Run")
+		}
+
+		// Find all the predecessor task runs
+		err = tx.Raw(`
+                SELECT pipeline_task_runs.* FROM pipeline_task_runs
+                LEFT JOIN pipeline_task_specs AS specs_right ON specs_right.id = pipeline_task_runs.pipeline_task_spec_id
+                LEFT JOIN pipeline_task_specs AS specs_left ON specs_left.id = specs_right.successor_id
+                LEFT JOIN pipeline_task_runs AS successors ON successors.pipeline_task_spec_id = specs_left.id
+                      AND successors.pipeline_run_id = pipeline_task_runs.pipeline_run_id
+                WHERE successors.id = ?`, ptRun.ID).Find(&predecessors).Error
+		if err != nil {
+			return errors.Wrap(err, "error finding task run predecessors")
+		}
+
+		// Get the job ID
+		var job struct{ ID int32 }
+		err = tx.Raw(`
+            SELECT jobs.id FROM pipeline_task_runs
+            LEFT JOIN pipeline_task_specs ON pipeline_task_specs.id = pipeline_task_runs.pipeline_task_spec_id
+            LEFT JOIN jobs ON jobs.pipeline_spec_id = pipeline_task_specs.pipeline_spec_id
+            WHERE pipeline_task_runs.id = ?
+        `, ptRun.ID).Scan(&job).Error
+		if err != nil {
+			return errors.Wrap(err, "error finding job ID")
+		}
+
+		// Call the callback
+		result := fn(job.ID, ptRun, predecessors)
+
+		// Update the task run record with the output and error
+		if result.Value != nil {
+			out := &JSONSerializable{Val: result.Value}
+			err = tx.Exec(`UPDATE pipeline_task_runs SET output = ?, error = NULL, finished_at = ? WHERE id = ?`, out, time.Now(), ptRun.ID).Error
+		} else if result.Error != nil {
+			logger.Errorw("Error in pipeline task", "error", result.Error)
+			errString := null.StringFrom(result.Error.Error())
+			err = tx.Exec(`UPDATE pipeline_task_runs SET output = NULL, error = ?, finished_at = ? WHERE id = ?`, errString, time.Now(), ptRun.ID).Error
+		}
+		if err != nil {
+			return errors.Wrap(err, "could not mark pipeline_task_run as finished")
+		}
+		return nil
+	})
+	return done, err
+}
+
+const (
+	postgresChannelRunStarted   = "pipeline_run_started"
+	postgresChannelRunCompleted = "pipeline_run_completed"
+)
+
+func (o *orm) ListenForNewRuns() (*utils.PostgresEventListener, error) {
+	listener := &utils.PostgresEventListener{
```

Comment:
I'm nervous about how many of these we are creating. Postgres connections are rather expensive.

Reply:
Let's profile

Reply:
That's already getting uncomfortable for me.

Most postgres installations have a hard max cap between 100-200 connections.

30 per node is a lot.

Reply:
Ah it won't be 30 per node, though.  Let's say a node op has a cluster of 3 nodes and 30 total jobs, that'd be roughly `3n + 30` -- the per-job connections aren't duplicated on each node.

Reply:
Still not great. We wind up in a situation where if a node operator adds more than say between 100-200 jobs suddenly their node will stop working with a bunch of cryptic "too many postgres connections" errors.